rfd/0124-plugin-secrets.md

```diff
@@ -0,0 +1,176 @@
+    api_token: example-token
+    basic_auth:
+      username: example-user
+      password: example-password
+    oauth_client_secret:
+      client_id: example-client-id
+      client_secret: example-client-secret
+```
+
+If a `Plugin` is using the oauth authorization code flow, however, it will not need any
+`PluginStaticCredentials` objects.
+
+### Changes to the `Plugin` object
+
+The `Plugin` object's `Credentials` field will have a new type of credential called
+`CredentialsRef` that contains a reference to a static plugin label. More than
+one static credential may match this label, in which case when starting the plugin, the most
+recently created credential will be used.
+
+```yaml
+credentials_ref:
+  labels:
+    # This label is the same label generated at plugin creation. This field will be
+    # injected by Teleport at creation.
+    teleport.internal/plugin: 48398071-9860-4be6-9f29-ca6ca5bc7155
+    env: prod
+    type: slack
+```
+
+### Access control
+
+`PluginStaticCredentials` can be written to and deleted by `RoleProxy`. At introduction,
+creating credentials is a flow that will only happen through the web UI, so only `RoleProxy`
+is needed. Future work may determine that we want to open up writing through the CLI.
+
+The object will be only readable by the `RoleAdmin` role (with one exception).
+This is because plugins run on the auth server, so it's the only role that needs to be
+able to actually read the static credentials.
+
+One `Plugin` object can refer to multiple static credentials and a `Plugin` can't refer
+to existing static credentials. Static credentials must be associated with one and only
+one plugin.
+
+The one exception here is Teleport Assist, which runs in the proxy. The proxy will be
+able to read plugin credentials with the name of `openai-default`, which is the name of
+the Assist plugin.
+
+### Lifecycle of the `PluginStaticCredentials` object
+
+#### Creation
+
+A `PluginStaticCredentials` object must be created at the same time that the `Plugin` is created.
+
+#### Deletion
+
+On deletion, the associated `Plugin` will be stopped and related `PluginStaticCredentials`
+objects will be deleted when deleting its associated plugin.
+
+### UX
+
+#### Web UI
+
+The web UI will remain largely the same with respect to entering credentials for specific
+plugins. This will largely depend on the specific application.
+
+### Audit events
+
+A number of new audit events will be created as part of this effort:
+
+| Event Name | Description |
+|------------|-------------|
+| `plugin.credentials.create` | Emitted when plugin credentials have been created. |
+| `plugin.credentials.delete` | Emitted when plugin credentials have been deleted. |
+
+### Security
+
+* Only the auth server is able to read plugin credentials. The proxy will be able to create
+them, and admins will be able to delete them.
+* The credentials will be stored in DynamoDB, which is encrypted at rest. This will provide
```

Comment:
Are these being stored in a dynamoDB table that is only used for these credentials?  Or what isolation do we have from other dynamoDB storage we are doing today?